src/sleep/sleep.c

```diff
@@ -263,40 +263,90 @@ static int execute(
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                /* Calculate current battery capacity and current time before suspension*/
+                last_capacity = read_battery_capacity();
+                if (last_capacity < 0)
+                        return log_error_errno(errno, "Error fetching battery capacity: %m");
 
-        r = timerfd_settime(tfd, 0, &ts, NULL);
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                before_timestamp = now(CLOCK_MONOTONIC);
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
+                previous_discharge_time = get_battery_discharge_rate();
+                if (previous_discharge_time < 0)
+                        log_error_errno(errno, "Error fetching battery discharge rate: %m");
+                else
+                        suspend_interval = ((previous_discharge_time * 60) - 30) * USEC_PER_MINUTE;
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s",
+                        FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_MONOTONIC);
+
+                        current_capacity = read_battery_capacity();
+                        if (current_capacity < 0)
+                                return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                        log_debug("%d%% battery discharged in %s time",
+                                        last_capacity - current_capacity,
+                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, USEC_PER_SEC));
+                        estimated_discharge_rate = (last_capacity - current_capacity) / (before_timestamp - after_timestamp);
```

Comment:
need to check that `before_timestamp - after_timestamp` is a positive value

Reply:
it should be `before_timestamp - after_timestamp` for it to be always positive. Mistakenly added opposite way.